### DISCUSSION TITLE
TIFF 32-bit floating values

### DISCUSSION THREAD

**Discussion by smnsndl** [ImageMagick/ImageMagick#2542] [Help] [answered]
Hello! 
I'm having issues with a 32-bit GeoTIFF DSM (Digital Surface Model), single channel with values.
This following example is just a mock-up regular tiff image that have the same problems with.
It was created using python and numpy random array full of float32 values
HeightxWidth 500x500, min/max values are -9999.93/9999.97

**Original mockup image visualized in QGIS**
![image10000-visualized-qgis](https://user-images.githubusercontent.com/9282309/92695724-2b218280-f349-11ea-857d-8fccbbef7916.png)

When I read this file and just try to write to a new tiff file all the negative values are cut off.
I've tried using -define quantum:format=floating-point and -depth 32 when writing to the new file but there is no difference in the result
**Maybe I'm doing something wrong and I just don't fully understand how the commands work yet.. hopefully someone can point me in the right direction.**
`magick .\image-10000.tiff -define quantum:format=floating-point -depth 32 .\image-10000-test.tiff`
I don't get any difference from just doing:
`magick .\image-100000.tiff .\image-10000-test.tiff`
![image10000-test-visualized-qgis](https://user-images.githubusercontent.com/9282309/92696492-4214a480-f34a-11ea-933f-fac8c5b465ac.png)

When doing the identify command on the original image it says the min value is: 1.17549e-38  (1.79369e-43) which is what the result image gets.
I've been searching around for days on forums and documentation and I can't seem to find a solution on how to keep the negative values.
Thanks, Simon

Info
[magick-identify-verbose.txt](https://github.com/ImageMagick/ImageMagick/files/5199928/magick-identify-verbose.txt)
[magick-version.txt](https://github.com/ImageMagick/ImageMagick/files/5199926/magick-version.txt)
[images.zip](https://github.com/ImageMagick/ImageMagick/files/5201115/images.zip)

**Comment by snibgo** [comment]
Your file image-10000.tiff works fine for me. With IM 7.0.8-64, with HDRI, these commands...
```
magick image-10000.tiff -define quantum:format=floating-point -depth 32 x.tiff

magick x.tiff -verbose info:
```
... has an output that includes:
```
  Channel statistics:
    Pixels: 250000
    Gray:
      min: -6.55345e+008  (-9999.93)
      max: 6.55348e+008 (9999.97)
      mean: 726246 (11.0818)
      standard deviation: 3.78606e+008 (5777.16)
      kurtosis: -1.20064
      skewness: -0.00165542
```

At the bottom of your magick-identify-verbose.txt, you have:
```
identify: UnableToOpenConfigureFile `colors.xml' @ warning/configure.c/GetConfigureOptions/702.

```
This suggests your installation is bad. If IM can't find its own XML files, something is wrong. I suggest you uninstall IM, and reinstall it.

**Reply by smnsndl** [reply]
OK, so I tried reinstalling the same version I had (7.0.10-29 Q16 x64 HDRI) and ran the same commands..
The colors.xml warning is gone but my issue remains.
I now have gone back to test the following versions:
- 7.0.9.9 Q16 X64 HDRI - same issue
- 7.0.9.0 Q16 X64 HDRI - same issue but I get min: 0 (0) instead of 1.17549e-38 (1.79369e-43)
- 7.0.8-9 Q16 x64 HDRI - **it works!**
Something seems to be wrong with the later versions...

**Comment by snibgo** [comment]
Oh, dear, yes, there is a bug. With v7.0.10-27, this command...
```
magick image-10000.tiff -define quantum:format=floating-point -depth 32 x.tiff
```
... makes x.tiff that contains only positive values. This is clearly a bug.

Investigating further, I find that this command works fine:
```
f:\web\im>%IMG701010%magick xc:Black -evaluate subtract 1000% txt:

# ImageMagick pixel enumeration: 1,1,65535,srgb
0,0: (-655360,-655360,-655360)  #000000000000  srgb(-1000.02%,-1000.02%,-1000.02%)
```

But these two commands don't work:
```
f:\web\im>%IMG701010%magick xc:Black -evaluate subtract 1000% -define quantum:format=floating-point -depth 32 x.tiff

f:\web\im>%IMG701010%magick x.tiff txt:
# ImageMagick pixel enumeration: 1,1,65535,gray
0,0: (1.17549e-38)  #000000000000  gray(4.57391e-41)
```
So the bug seems to be when writing the tiff file. The problem also occurs with MIFF format.

A developer may comment.

**Comment by smnsndl** [comment]
I thought this version (7.0.8-9 Q16 x64 HDRI) would work with another 32-bit image but apparently...
```
magick dsm.tif -verbose info:
magick.exe: Unknown field with tag 33550 (0x830e) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/997.
magick.exe: Unknown field with tag 33922 (0x8482) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/997.
magick.exe: Unknown field with tag 34735 (0x87af) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/997.
magick.exe: Unknown field with tag 34736 (0x87b0) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/997.
magick.exe: Unknown field with tag 34737 (0x87b1) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/997.
magick.exe: Unknown field with tag 42113 (0xa481) encountered. `TIFFReadDirectory' @ warning/tiff.c/TIFFWarnings/997.
magick.exe: Sorry, can not handle images with 32-bit samples. `dsm.tif' @ error/tiff.c/TIFFErrors/652.
```
I'll check and see if I can share this image and get back if anybody wants to take a look

**Reply by smnsndl** [reply]
Sorry for taking my sweet time.
Can anyone tell me why does image can't be handled?
[dsm.zip](https://github.com/ImageMagick/ImageMagick/files/5218108/dsm.zip)

**Comment by dlemstra** [comment, marked as answer]
The issue seemed to be an invalid check for the maximum negative value when decoding the TIFF image. A patch has just been pushed to resolve that issue and this will be resolved in the next release:

```
I:\issues\im7\2542>magick xc:Black -evaluate subtract 1000% -define quantum:format=floating-point -depth 32 x.tiff
I:\issues\im7\2542\magick x.tiff txt:
# ImageMagick pixel enumeration: 1,1,65535,gray
0,0: (-655360)  #000000000000  gray(-1000.02%)
```

**Comment by snibgo** [comment]
> Can anyone tell me why does image can't be handled? dsm.zip

What do you mean "can't be handled"? Most of the values are outside 0 to 100% of QuantumRange. With IM v7.0.10-27:
```
magick dsm.tif -auto-level x.png
```
Then x.png shows an image. (But I have no idea what the image is.)

But you might wait for the version that contains dlemstra's patch.

**Reply by smnsndl** [reply]
With version 7.0.8-9 Q16 x64 HDRI magick literally says..
"magick.exe: Sorry, can not handle images with 32-bit samples. `dsm.tif' @ error/tiff.c/TIFFErrors/652"
I can't even do the command you wrote, is it a version issue?
The image is Digital Surface Model (DSM) over a lake+some forest, the values are height values of that place.

Maybe I have misinterpreted the way this image is structured or perhaps the software rounds off some values. How did you check that the values are outside quantumrange?
Thanks

**Comment by snibgo** [comment]
Yes, it is a version problem. IM processing of TIFF 32-bit floats is currently unstable, so I suggest you wait for the version that contains dlemstra's patch.

With v7.0.8-64, "magick dsm.tif -verbose info:" says:
```
  Channel statistics:
    Pixels: 1093345
    Gray:
      min: -6.55284e+008  (-9999)
      max: 8.01362e+006 (122.28)
      mean: 5.36513e+006 (81.8667)
      standard deviation: 2.05289e+007 (313.251)
      kurtosis: 1029.99
      skewness: -32.0999
      entropy: 0.0145885
```
So after reading the "image" into memory, values range from -6.6e+8 to +8.01e+6. My IM is Q16, so this is in a nominal scale of 0 to 65535. The numbers in parentheses are on a scale of 0.0 to 1.0, so are (probably) the values actually found in the file: -9999 to +122.28.

These might represent heights in metres (or feet or inches ...) below or above a geoid. Documentation from the source of the file should explain.

EDIT to add: "exiftool gsm.tif" gives useful metadata. The values seem to be in metres above WGS84, from -401 to +122 metres, with "-9999" meaning "no data".
